Initial working appscale-agents from appscale-tools repo

scragraham opened the pull request:
This is the initial PR for the working version of appscale.agents.
 - Makefile helps in installing, building, and version bumping
   - make upload is not implemented.
   - The makefile tries to ensure that a virtualenv is active before running certain commands.

Running 'make' without any arguments will print the help:
```
$ make
Testing
test       - Run Python Unit tests 

Installation
install    - Install via pip, ensuring a virtualenv
install-no-venv - Install via pip without ensuring a virtualenv

Build (for PyPi)
build      - Build distro for pypi upload
clean      - Clean build directories
upload     - Upload distribution to pypi

Utilities
bump-major - Bump major version number for appscale-agents
bump-minor - Bump minor version number for appscale-agents
bump-patch - Bump patch version number for appscale-agents
```

 - config.AppScaleState is a minimal copy of appscale.tools.local_state
 - appscale_logger is a direct pull from appscale.tools
 - agents have been converted to not rely on appscale.tools at all
 - copied over util/bump_version.sh from appscale-tools
 - Unit tests are passing, I suspect there needs to be some cleanup.
 - I tried to keep changes to a minimum. (spacing from 2 to 4, some minor refactoring)
   - I will format the files in a future PR to avoid too many changes.
 - requests moved to 2.20.0 to address CVE security issue.

scragraham committed:
working appscale.agents split off from appscale-tools

scragraham committed:
requests to version 2.20.0

scragraham committed:
install flexmock before running tests

scragraham committed:
- added .gitignore with typical settings.
- modified Makefile, added, clean/build/upload.
- fixed test_upgrade_json to properly pass the test.

scragraham committed:
change public ip to a private network ip

scragraham committed:
makefile adjustment, fix test_upgrade_json expected results

cdonati commented on appscale/agents/azure_agent.py:
```diff
@@ -72,8 +72,9 @@
 # AppScale-specific imports
-from appscale.tools.appscale_logger import AppScaleLogger
-from appscale.tools.local_state import LocalState
+from config import AppScaleState
```
Use either absolute imports (eg. `from appscale.agents.config`) or explicit relative imports (eg. `from .config`)

cdonati commented on appscale/agents/appscale_logger.py:
```diff
@@ -0,0 +1,106 @@
+        printed to stdout.
+        """
+        if is_verbose:
+            print message
```
The `print()` function would be nicer here since it's forward-compatible.

cdonati commented on appscale/agents/base_agent.py:
```diff
@@ -249,12 +249,7 @@ def diff(self, list1, list2):
-      if item not in list2:
-        diffed_list.append(item)
-    return diffed_list
+    return list(set(list1).difference(set(list2)))
```
Not sure if it's important, but the previous method preserved the order of items from `list1`.

scragraham committed:
fix import of config module

scragraham committed:
adjust packages

scragraham committed:
use correct variable name

scragraham committed:
missed one LocalState reference

scragraham committed:
revert base_agent diff() to original version

scragraham committed:
use print() function instead of print builtin

scragraham committed:
add venv target in makefile

scragraham committed:
added cleanup of venv dir, minor changes to virtualenv recipe

scragraham commented:
Unit tests:

```
make test                                                     ✔
Collecting flexmock
Installing collected packages: flexmock
Successfully installed flexmock-0.10.2
test_config_path (test_config_appscale_state.TestAppScaleState) ... ok
test_generate_rsa_key (test_config_appscale_state.TestAppScaleState) ... ok
test_get_app_id (test_config_appscale_state.TestAppScaleState) ... ok
test_get_app_secret_key (test_config_appscale_state.TestAppScaleState) ... ok
test_get_client_secrets_location (test_config_appscale_state.TestAppScaleState) ... ok
test_get_group (test_config_appscale_state.TestAppScaleState) ... ok
test_get_infrastructure_option (test_config_appscale_state.TestAppScaleState) ... ok
test_get_oauth2_storage_location (test_config_appscale_state.TestAppScaleState) ... ok
test_get_project (test_config_appscale_state.TestAppScaleState) ... ok
test_get_resource_group (test_config_appscale_state.TestAppScaleState) ... ok
test_get_storage_account (test_config_appscale_state.TestAppScaleState) ... ok
test_get_subscription_id (test_config_appscale_state.TestAppScaleState) ... ok
test_get_tenant_id (test_config_appscale_state.TestAppScaleState) ... ok
test_get_zone (test_config_appscale_state.TestAppScaleState) ... ok
test_locations_json_location (test_config_appscale_state.TestAppScaleState) ... ok
test_locations_yaml_location (test_config_appscale_state.TestAppScaleState) ... ok
test_private_key (test_config_appscale_state.TestAppScaleState) ... ok
test_public_key (test_config_appscale_state.TestAppScaleState) ... ok
test_shell_exceptions (test_config_appscale_state.TestAppScaleState) ... ok
test_ssh_key (test_config_appscale_state.TestAppScaleState) ... ok
test_upgrade_json_file (test_config_appscale_state.TestAppScaleState) ... ok
test_write_key_file (test_config_appscale_state.TestAppScaleState) ... ok

test_create_agent_azure (test_agent_factory.TestAgentFactory) ... ok
test_create_agent_ec2 (test_agent_factory.TestAgentFactory) ... ok
test_create_agent_euca (test_agent_factory.TestAgentFactory) ... ok
test_create_agent_gce (test_agent_factory.TestAgentFactory) ... ok
test_create_agent_openstack (test_agent_factory.TestAgentFactory) ... ok
test_create_invalid_infrastructure (test_agent_factory.TestAgentFactory) ... ok

test_euca_run_instances (test_agent_euca.TestEucaAgent) ... ok
test_ec2_run_instances (test_agent_ec2.TestEC2Agent) ... ok
test_ec2_run_instances_agentruntimeexception (test_agent_ec2.TestEC2Agent) ... ok
test_diff_helper (test_base_agent.TestBaseAgent) ... ok
test_gce_run_instances (test_agent_gce.TestGCEAgent) ... ok

----------------------------------------------------------------------
Ran 33 tests in 80.398s

OK
```

scragraham committed:
Removal of AppScaleLogger in appscale.agents
 - Since this is a library now, I removed the AppScaleLogger
 - All logging is done via std python logging mechanisms.
 - Current Loggers:
   - appscale.agents - top level logger
   - appscale.agents.{gce|euca|ec2|azure}_agent - Cloud specific
     agent logger.

whoarethebritons commented on appscale/agents/azure_agent.py:
```diff
@@ -1610,16 +1605,15 @@ def get_cloud_params(self, keyname):
+      self.PARAM_ZONE: AppScaleState.get_zone(keyname),
+      self.PARAM_SUBSCRIBER_ID: AppScaleState.get_subscription_id(keyname),
+      self.PARAM_APP_ID: AppScaleState.get_app_id(keyname),
+      self.PARAM_APP_SECRET: AppScaleStatee.get_app_secret_key(keyname),
```
"AppScaleStatee"

scragraham replied:
fixed

whoarethebritons commented on appscale/agents/config.py:
```diff
@@ -0,0 +1,400 @@
+        file (used to interact with Google Compute Engine) can be found.
+        
+        Args:
+        keyname: A str representing the SSH keypair name used for this AppScale
```
indentation

scragraham replied:
fixed

scragraham committed:
- Fix object name typo
 - Fix indentation on doc strings
